src/passes/OptimizeInstructions.cpp

```diff
@@ -1929,23 +1929,41 @@ struct OptimizeInstructions
     Builder builder(*getModule());
     auto nullType = curr->type.getHeapType().getBottom();
     {
-      auto* ref = curr->ref;
+      auto** refp = &curr->ref;
       while (1) {
+        auto* ref = *refp;
+
         auto result = GCTypeUtils::evaluateCastCheck(ref->type, curr->type);
 
         if (result == GCTypeUtils::Success) {
-          // The cast will succeed, but we can't just remove the cast and
-          // replace it with `ref` because the intermediate expressions might
-          // have had side effects. We can replace the cast with a drop followed
-          // by a direct return of the value, though.
-          //
-          // TODO: Do this for non-null values as well by storing the value to
-          // return in a tee.
+          // The cast will succeed. If there were no intermediate expressions,
```

Comment:
```suggestion
          // The cast will succeed. This can only happen if the ref is a subtype of
          // the cast instruction, which means we can replace the cast with the ref.
          assert(Type::isSubType(ref->type, cast->type));
          // If there were no intermediate expressions,
```
(without this, it's not obvious why these replacements are all valid)